QuizApp/QuizApp/UI/Screens/SolvingQuiz/SolvingQuizViewModel.swift

```diff
@@ -28,24 +32,59 @@ class SolvingQuizViewModel {
                 progressColors = [UIColor](repeating: unansweredColor, count: quiz.questions.count)
                 progressColors[0] = .white
             } catch {
-
+                errorMessage = """
+Couldn't start quiz.
+Please try again.
+"""
             }
         }
     }
 
     @MainActor
     func selectedAnswer(with id: Int) {
-        guard let questionIndex = progressColors.firstIndex(of: .white) else { return }
+        if isQuestionAnswered(for: id) { return }
+
+        let question = quiz.questions[currentQuestionIndex]
+
+        changeAnswerColors(for: question, selectedAnswerId: id)
 
-        let isAnswerCorrect = quiz.questions[questionIndex].correctAnswerId == id
+        let isAnswerCorrect = quiz.questions[currentQuestionIndex].correctAnswerId == id
         if isAnswerCorrect {
-            progressColors[questionIndex] = .correctAnswerColor
+            progressColors[currentQuestionIndex] = .correctAnswerColor
+            correctQuestions += 1
         } else {
-            progressColors[questionIndex] = .incorrectAnswerColor
+            progressColors[currentQuestionIndex] = .incorrectAnswerColor
+        }
+
+        if currentQuestionIndex < progressColors.count - 1 {
+            progressColors[currentQuestionIndex + 1] = .white
+            currentQuestionIndex += 1
+        } else {
+            finishQuiz()
+        }
+    }
+
+    private func isQuestionAnswered(for id: Int) -> Bool {
+        for index in 0..<currentQuestionIndex {
+            if quiz.questions[index].answers.contains(where: { $0.id == id }) {
+                return true
+            }
         }
 
-        if questionIndex + 1 < progressColors.count {
-            progressColors[questionIndex + 1] = .white
+        return false
+    }
+
+    private func changeAnswerColors(for question: Question, selectedAnswerId: Int) {
+        quiz = QuizStartResponse(quiz, id: question.id, selectedAnswerId: selectedAnswerId)
+    }
+
+    private func finishQuiz() {
+        let delayInMillis = 400
+        DispatchQueue.main.asyncAfter(deadline: .now() + .milliseconds(delayInMillis)) { [weak self] in
+            guard let self = self else { return }
+
+            let text = "\(self.correctQuestions)/\(self.quiz.questions.count)"
+            self.router.showResults(with: text)
```

Comment:
Do not decide here what should be drawn in some other VC. Instead, pass variables to it.